offchain/danaswap-api/src/DanaSwap/Api.purs

```diff
@@ -9,30 +17,119 @@ import Contract.Prelude
 import Contract.Address (getWalletAddress, getWalletCollateral, scriptHashAddress)
 import Contract.Log (logDebug', logInfo')
 import Contract.Monad (Contract, liftContractM)
-import Contract.PlutusData (Datum(..), PlutusData(..), toData)
+import Contract.PlutusData (Datum(..), PlutusData(..), Redeemer(..), toData)
+import Contract.Prim.ByteArray (hexToByteArray)
 import Contract.ScriptLookups as Lookups
 import Contract.Scripts (MintingPolicy, Validator, mintingPolicyHash, validatorHash)
-import Contract.Transaction (TransactionInput)
+import Contract.Transaction (TransactionInput, TransactionOutputWithRefScript)
 import Contract.TxConstraints (DatumPresence(..), TxConstraints)
 import Contract.TxConstraints as Constraints
 import Contract.Utxos (getUtxo)
-import Contract.Value (CurrencySymbol, adaToken, mpsSymbol, scriptCurrencySymbol)
+import Contract.Value (CurrencySymbol, TokenName, adaToken, mkTokenName, mpsSymbol, scriptCurrencySymbol, symbols, valueOf)
 import Contract.Value as Value
 import Ctl.Util (buildBalanceSignAndSubmitTx, getUtxos, waitForTx)
 import DanaSwap.CborTyped (configAddressValidator, liqudityTokenMintingPolicy, poolAddressValidator, poolIdTokenMintingPolicy, simpleNft)
 import Data.BigInt as BigInt
 import Data.List (head)
-import Data.Map (keys)
+import Data.Map (Map, keys)
 import Data.Map as Map
 import Data.Set (toUnfoldable)
+import Effect.Exception (throw)
 
 type Protocol =
   { configUtxo :: TransactionInput
-  , poolVal :: Validator
+  , poolAdrVal :: Validator
   , liquidityMP :: MintingPolicy
   , poolIdMP :: MintingPolicy
   }
 
+type PoolId = TokenName
+
+getAllPools :: Protocol -> Contract () (Map TransactionInput TransactionOutputWithRefScript)
+getAllPools protocol@{ poolAdrVal } =
+  getUtxos (scriptHashAddress $ validatorHash poolAdrVal)
+    <#> Map.filter (hasNft protocol)
+
+getPoolById :: Protocol -> PoolId -> Contract () (TransactionInput /\ TransactionOutputWithRefScript)
+getPoolById protocol@{ poolIdMP } token = do
+  pools <- getAllPools protocol
+  cs <- liftContractM "Failed to get the currency symbol for the protocols mintingPolicy" $ mpsSymbol $ mintingPolicyHash poolIdMP
+  let valid = Map.filter (\vault -> valueOf (unwrap (unwrap vault).output).amount cs token > BigInt.fromInt 0) pools
+  case Map.toUnfoldableUnordered valid of
+    [] -> liftEffect $ throw "no pools with that ID"
+    [ vault ] -> pure vault
+    _ -> liftEffect $ throw "more than one pool with the same ID, this is really bad"
```

Comment:
I believe the DanaSwap Stats data (the list of pools) could reduce the complexity of this request.

cc @CSVdB

Reply:
I kinda doubt that'll be simpler than this tbh.